core-metastore/src/main/java/org/zalando/nakadi/repository/db/SubscriptionDbRepository.java

```diff
@@ -121,8 +125,88 @@ public void deleteSubscription(final String id)
         }
     }
 
+    public static enum TokenType {
+        FORWARD('F'),
+        BACKWARD('B'),
+        BACKWARD_INCL('I'),
+        ;
+        private final char symbol;
+
+        TokenType(char symbol) {
+            this.symbol = symbol;
+        }
+
+        public static TokenType of(final char symbol) {
+            for (final TokenType tt : TokenType.values()) {
+                if (tt.symbol == symbol) {
+                    return tt;
+                }
+            }
+            throw new IllegalArgumentException("Token type not supported");
+        }
+    }
+
+    public static class Token {
+        private final String subscriptionId;
+        private final TokenType tokenType;
+
+        public Token(final String subscriptionId, final TokenType tokenType) {
+            this.subscriptionId = subscriptionId;
+            this.tokenType = tokenType;
+        }
+
+        public boolean hasSubscriptionId() {
+            return !StringUtils.isEmpty(subscriptionId);
+        }
+
+        public String getSubscriptionId() {
+            return subscriptionId;
+        }
+
+        public TokenType getTokenType() {
+            return tokenType;
+        }
+
+        public static Token createEmpty() {
+            return new Token("", TokenType.FORWARD);
+        }
+
+        public String encode() {
+            return tokenType.symbol + subscriptionId;
+        }
+
+        public static Token parse(final String value) {
+            if (StringUtils.isEmpty(value) || value.length() < 1) {
+                return createEmpty();
+            }
+            return new Token(value.substring(1), TokenType.of(value.charAt(0)));
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+            Token token = (Token) o;
+            return Objects.equals(subscriptionId, token.subscriptionId) &&
+                    tokenType == token.tokenType;
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(subscriptionId, tokenType);
+        }
+
+        @Override
+        public String toString() {
+            return "Token{" +
+                    "subscriptionId='" + subscriptionId + '\'' +
+                    ", tokenType=" + tokenType +
+                    '}';
+        }
+    }
+
     public List<Subscription> listSubscriptions(final Set<String> eventTypes, final Optional<String> owningApplication,
-                                                final int offset, final int limit)
+                                                final int offset, final int limit, @Nullable final Token token)
```

Comment:
This function is very long and complex by now. 
It will be nice if we can extract a `QueryBuilder` class with the `StringBuilder queryBuilder`, `List<String> clauses`, and `List<Object> params`.

Reply:
agree, changed to another approach